gnome-image-installer/util/eos-reformatter.c

```diff
@@ -0,0 +1,972 @@
+    }
+
+  /* A bit awkwardly, exhaust doesn't apply when we create new buffers */
+  if (reformatter->used_buffers < reformatter->total_buffers)
+    {
+      /* No free buffers, but we can create one */
+      gsize offset = reformatter->used_buffers * (reformatter->buffer_size + reformatter->page_size);
+      guchar *unaligned = reformatter->pool + offset;
+      gsize page_size = reformatter->page_size;
+
+      buf = g_new0(EosAlignedBuffer, 1);
+      buf->ptr = (guchar*) (((gintptr) (unaligned + page_size)) & (~(page_size - 1)));
+      buf->len = 0;
+      reformatter->used_buffers++;
+    }
+  else
+    {
+      /* If we can't create a new buffer, we need to wait for one */
+      EosAlignedBuffer *buf;
+
+      while (!exhaust && g_async_queue_length (reformatter->free_queue) < half_buffers);
+        {
+          g_thread_yield();
+        }
+
+      do
+        {
+          g_thread_yield();
+          buf = g_async_queue_try_pop (reformatter->free_queue);
+        }
+      while (buf == NULL && reformatter->error == NULL);
+      buf->len = 0;
+      return buf;
+    }
+
+  return buf;
+}
+
+static gint
+eos_reformatter_prepare_read (EosReformatter *reformatter)
+{
+  GFileInfo *info;
+  GFile *file = g_file_new_for_path (reformatter->image);
+  GError *error = NULL;
+
+  info = g_file_query_info (file,
+                            G_FILE_ATTRIBUTE_STANDARD_CONTENT_TYPE ","
+                            G_FILE_ATTRIBUTE_STANDARD_SIZE,
+                            G_FILE_QUERY_INFO_NONE,
+                            NULL,
+                            &error);
+
+  if (info == NULL)
+    {
+      reformatter->error = g_error_new (EOS_REFORMATTER_ERROR, 0, _("Internal error"));
+      g_error_free (error);
+    }
+  else
+    {
+      const gchar *type = NULL;
+
+      /* If we weren't seeded with the target size, use the file size.
+       * This is of course wrong for compressed files */
+      if (reformatter->write_total_bytes == 0)
+        {
+          reformatter->write_total_bytes = g_file_info_get_size (info);
+        }
+
+      type = g_file_info_get_content_type (info);
+      if (g_str_has_suffix(type, "-xz-compressed"))
+        {
+          reformatter->decompressor = G_CONVERTER (gdu_xz_decompressor_new ());
+        }
+      else if (g_str_equal(type, "application/gzip"))
+        {
+          reformatter->decompressor = G_CONVERTER (g_zlib_decompressor_new (G_ZLIB_COMPRESSOR_FORMAT_GZIP));
+        }
+  }
+
+  g_object_unref (file);
+
+  if (info != NULL)
+    g_object_unref (info);
+
+  return open(reformatter->image, O_RDONLY);
+}
+
+static gboolean
+eos_reformatter_read (EosReformatter *reformatter, GAsyncQueue *outq)
+{
+  gboolean ret = TRUE;
+  guint64 len = 0;
+  EosAlignedBuffer *buf = eos_reformatter_get_free_buffer (reformatter, FALSE);
+
+  if (reformatter->error != NULL)
+    {
+      /* Error is set so send EOS and abort */
+      if (buf != NULL)
```

Comment:
Surely `buf == NULL` only if an error has occurred, so you can just write a `NULL` (say) to the `outq` and return.